host-go/runtimes/wazero/runtime.go

```diff
@@ -120,8 +135,12 @@ func (m *wModule) NewInstance(functionName string, paramSets ...map[string]any)
 			}
 			return module.MemSize(r[0]), nil
 		},
-		Transform: func(u module.MemSize) (module.MemSize, error) {
-			r, err := transform.Call(ctx, uint64(u))
+		Transform: func(next func() module.MemSize) (module.MemSize, error) {
+			// By assigning the next function immediately prior to calling transform, we allow multiple
+			// pipeline stages to share the same wasm instance - provided they are not called concurrently.
```

Comment:
thought: this line may answer my above question. Specifically "not called concurrently". Does this seem like a limiter here, theres not much else in the codebase that im aware of that would limit concurrency of multiple instances besides this. Thoughts?

Reply:
Enumerables in general do not work concurrently, as they inherently rely on state mutation/progression.

If we wanted to allow the/any enumerable instance to be thread safe we'd be basically wrapping the public functions in locks anyway.

E.g. two different threads can never process a `next` call on the same enumerable concurrently, unless you deliberate chose to allow the `next` calls to return two different, divergent, enumerable instances.

Reply:
My question was more about multiple concurrent instances on a single runtime. Obviously we have the ability to instanciate multiple instances from the same cache to avoid additional complition. But if the `next` func is registered to the runtime and not the instance, than cant only a single instance run at a time, despite using different source enumerables?

Reply:
Ah okay, I understand :)

> But if the next func is registered to the runtime and not the instance, than cant only a single instance run at a time, despite using different source enumerables?

The next function is registered to the instance, not the runtime.  In wasmer and wasmtime the runtime is shared across instances, but `next` is instance specific.  In wazero each instance has its own runtime (so only one `next` func per instance).

There are a lot of tests that test multiple lens instances from a single lens runtime.  I can't imagine them passing if the `next` functions were coupled to the runtime and not the instance.

Reply:
ooh, didnt realize wazero each instance has its own runtime, then the compilation cache is used across runtimes. Gotcha. 

I think we can further optimize by using 1 shared for all instances but likely not till we figure out the import/instance stuff. So this is a great work around for now.

Cheers.